torch_xla/csrc/runtime/stablehlo_composite_helper.cc

```diff
@@ -0,0 +1,391 @@
+#include "torch_xla/csrc/runtime/stablehlo_composite_helper.h"
+
+#include <limits>
+#include <string>
+#include <tuple>
+#include <unordered_map>
+
+#include "absl/log/log.h"
+#include "absl/strings/str_cat.h"
+#include "mlir/IR/IRMapping.h"
+#include "single_include/nlohmann/json.hpp"
+#include "stablehlo/dialect/StablehloOps.h"
+
+namespace torch_xla {
+namespace runtime {
+
+namespace {
+
+using nlohmann::json;
+
+static bool IsXlaMarkTensorOp(mlir::Operation* op) {
+  if (op == nullptr) {
+    return false;
+  }
+  if (op->getNumOperands() != 1 || op->getNumResults() != 1) {
+    return false;
+  }
+  if (!llvm::isa<mlir::stablehlo::CustomCallOp>(op)) {
+    return false;
+  }
+  auto target_name =
+      op->getAttr("call_target_name").dyn_cast<mlir::StringAttr>();
+  if (target_name == nullptr || target_name.str() != "xla_mark_tensor") {
+    return false;
+  }
+  return true;
+}
+
+struct BoundaryMetadata {
+  std::string name;
+  int64_t id;
+  int64_t pos;
+  bool is_input;
+  std::unordered_map<std::string, json> attrs;
+
+  std::string boundary_id() const { return absl::StrCat(name, "__", id); }
+
+  auto uid() const { return std::forward_as_tuple(name, id, pos, is_input); }
+
+  bool operator==(const BoundaryMetadata& other) const {
+    return uid() == other.uid();
+  }
+  bool operator<(const BoundaryMetadata& other) const {
+    return uid() < other.uid();
+  }
+
+  static std::unique_ptr<BoundaryMetadata> Parse(llvm::StringRef str) {
+    auto j = json::parse(str, /*cb=*/nullptr, /*allow_exceptions=*/false);
+    return Build(j);
+  }
+
+ private:
+  static std::unique_ptr<BoundaryMetadata> Build(
+      const nlohmann::basic_json<>& j) {
+    BoundaryMetadata metadata;
+#define FIELD(key, value_type)                           \
```

Comment:
I would avoid macro is possible since it makes compilation error hard to read, is there an easy way to just make it a helper function?

Reply:
Done. PTAL.